Packages/com.unity.inputsystem/InputSystem/Devices/FakeScreenKeyboard.cs

```diff
@@ -0,0 +1,132 @@
+                    ms_Instance = new FakeScreenKeyboard();
+                return ms_Instance;
+            }
+        }
+
+        string m_InputFieldText;
+        RangeInt m_Selection;
+        class FakeScreenKeyboardDispatcher : MonoBehaviour
+        {
+        }
+
+        private FakeScreenKeyboardDispatcher Dispatcher
+        {
+            get
+            {
+                var go = GameObject.Find(nameof(FakeScreenKeyboard));
+                if (go == null)
+                {
+                    go = new GameObject(nameof(FakeScreenKeyboard));
+                    go.AddComponent<FakeScreenKeyboardDispatcher>();
+                }
+                return go.GetComponent<FakeScreenKeyboardDispatcher>();
+            }
+        }
+
+        protected override void InternalShow()
+        {
+            m_InputFieldText = m_ShowParams.initialText;
+            m_Selection = new RangeInt(m_InputFieldText.Length, 0);
+            // Delay keyboard show
+            Dispatcher.StartCoroutine(QueueStatusChangeVisible());
+        }
+
+        IEnumerator Waiting()
+        {
+#if UNITY_EDITOR
+            // WaitForEndOfFrame doesn't work in batch mode
+            int startFrame = Time.frameCount;
+            return new WaitUntil(() => Time.frameCount - startFrame >= 1);
+#else
+            yield return new WaitForEndOfFrame();
+#endif
+        }
+
+        private IEnumerator QueueStatusChangeVisible()
+        {
+            yield return Waiting();
+            ReportStateChange(ScreenKeyboardState.Visible);
+        }
+
+        protected override void InternalHide()
+        {
+            // Delay keyboard hide
+            Dispatcher.StartCoroutine(QueueStatusChangeDone());
+        }
+
+        private IEnumerator QueueStatusChangeDone()
+        {
+            yield return Waiting();
+            ReportStateChange(ScreenKeyboardState.Done);
+        }
+
+        private IEnumerator QueueStatusChangeCancel()
+        {
+            yield return Waiting();
+            ReportStateChange(ScreenKeyboardState.Canceled);
+        }
+
+        private bool IsSelectionEqual(RangeInt a, RangeInt b)
+        {
+            return a.end == b.end && a.start == b.start;
+        }
+
+        private void OnSelectionChange(RangeInt newSelection)
+        {
+            if (IsSelectionEqual(newSelection, m_Selection))
+                return;
+            m_Selection = newSelection;
+            ReportSelectionChange(m_Selection.start, m_Selection.length);
+        }
+
+        public override string inputFieldText
+        {
+            get => m_InputFieldText;
+            set
+            {
+                if (m_InputFieldText.Equals(value))
+                    return;
+                m_InputFieldText = value;
+                // Note: Order is important, the selection is reported first
+                OnSelectionChange(new RangeInt(m_InputFieldText.Length, 0));
+                ReportInputFieldChange(value);
```

Comment:
yes, invoking property, would invoke the event

>With later choice you have a before/after problem: is event fired before the change has occurred or after?
after of course, then if in callback you query text property it would contain changed value.

>Consider the scenario where changed event does modify the value in some way and then sets the property. In current implementation it would lead to cascading event firing with potential of infinite recursion.
the callback wouldn't be called if you call property but provide same value. Obviously if you code your callback in a wrong way, like modifying text by adding some letter, this would cause callback to be infinetly called, but that's a legit error from user, which I think would be expected.

> Make events non-programmatic
sadly this is not possible, for ex., in Android case native onSelectionChanged callback doesn't differentiate between UI change or programmatic change. The same is with setText. It might be possible on IOS I think...

> Single handler only event implementation and new value being passed as ref to be modifiable in an event
not good, if event are coming as dispatched calls (I think that's the case on WSA), where events are called on another thread, and we'll be forced to dispatch it on main thread, so you can't change the value you receive. This also doesn't sound very safe. Unless I am misunderstanding.



Again, I don't think this is such a big problem, since like I said, the callback is called only in case where VALUE CHANGES.

Reply:
Resolved in dicussion